torch/package/package_exporter.py

```diff
@@ -588,6 +709,112 @@ def _write(self, filename, str_or_bytes):
+                if attrs.get("action") == _ModuleProviderAction.INTERN:
+                    interns.add(module_name)
+
+            include_filter = interns.copy()
+            for intern in interns:
+                for dep in self.dependency_graph.successors(intern):
+                    include_filter.add(dep)
+
+            raise PackagingError(denied, broken, unhandled, include_filter)
+
+        for pattern, _, allow_empty in self.patterns:
+            if not allow_empty and pattern not in self.matched_patterns:
+                raise EmptyMatchError(
+                    f"Exporter did not match any modules to {pattern}, which was marked as allow_empty=False"
+                )
+
+    def _execute_dependency_graph(self):
+        """Takes a finalized dependency graph describing how to package all
+        modules and executes it, writing to the ZIP archive.
+        """
+        self._validate_dependency_graph()
+
+        extern_modules = []
+        _mock_written = False
+        for module_name, attrs in self.dependency_graph.nodes.items():
+            action = attrs["action"]
+
+            if action == _ModuleProviderAction.EXTERN:
+                for hook in self._extern_hooks.values():
+                    hook(self, module_name)
+
+                extern_modules.append(module_name)
+
+            elif action == _ModuleProviderAction.MOCK:
+                for hook in self._mock_hooks.values():
+                    hook(self, module_name)
+
+                if not _mock_written:
+                    mock_file = str(Path(__file__).parent / "_mock.py")
+                    self._write_source_string(
+                        "_mock", _read_file(mock_file), is_package=False
+                    )
+                    _mock_written = True
+
+                is_package = hasattr(self._import_module(module_name), "__path__")
+                self._write_source_string(module_name, _MOCK_IMPL, is_package)
+
+            elif action == _ModuleProviderAction.INTERN:
+                for hook in self._intern_hooks.values():
+                    hook(self, module_name)
+
+                # The node in the dependency graph contains metadata that tells us
+                # how to intern the module.
+                if "provided" not in attrs:
+                    raise AssertionError(
+                        f"Module was marked `intern` but not provided: {module_name}"
+                    )
+
+                if attrs.get("is_pickle") is True:
+                    # This node came from save_source_pickle, we don't need to write any source for it.
+                    continue
+
+                is_package = attrs["is_package"]
+                source = attrs["source"]
+                self._write_source_string(module_name, source, is_package)
+
+            else:
+                raise AssertionError(
+                    f"Invalid action: {module_name}, {action}. Please report a bug to PyTorch."
+                )
```

Comment:
I think it is unintuitive that nodes that matched some deny pattern are present in the graph with `action = DENY`, and that such graphs are expected to be detected in `_validate_dependency_graph`.

`extern`, `intern`, and `mock` all convey information about how the dependency graph should be executed (i.e. do you pull in the deps of a given node or not) but deny does not.

Reply:
The reason for putting them in the graph is mostly so that we can continue on and present all errors to the user. We should never pass validation if there are denied modules in the graph though.